internal/drain_runner/runner.go

```
@@ -0,0 +1,131 @@
+			continue
+		}
+		if !done {
+			runner.logger.Info("preprocessor still pending", "node_name", candidate.Name, "preprocessor", pre.GetName())
+			allPreprocessorsDone = false
+		}
+	}
+
+	if !allPreprocessorsDone {
+		runner.logger.Info("waiting for preprocessors to be done before draining", "node_name", candidate.Name)
+		return nil
+	}
+
+	runner.logger.Info("all preprocessors of candidate are done; will start draining", "node_name", candidate.Name)
+	return runner.drain(ctx, candidate)
+}
+
+func (runner *drainRunner) drain(ctx context.Context, candidate *corev1.Node) error {
+	candidate, err := k8sclient.TaintNode(ctx, runner.client, candidate, runner.clock.Now(), k8sclient.TaintDraining)
+	if err != nil {
+		return err
+	}
+
+	err = runner.drainer.Drain(ctx, candidate)
+	if err != nil {
+		return runner.removeFailedCandidate(ctx, candidate, err.Error())
+	}
+
+	_, err = k8sclient.TaintNode(ctx, runner.client, candidate, runner.clock.Now(), k8sclient.TaintDrained)
+	return err
+}
+
+func (runner *drainRunner) removeFailedCandidate(ctx context.Context, candidate *corev1.Node, reason string) error {
+	err := runner.retryWall.SetNewRetryWallTimestamp(ctx, candidate, reason, runner.clock.Now())
+	if err != nil {
+		return err
+	}
+
+	_, err = k8sclient.UntaintNode(ctx, runner.client, candidate)
+	return err
+}
+
+func (runner *drainRunner) getDrainCandidates(ctx context.Context, key groups.GroupKey) ([]*corev1.Node, error) {
+	nodes, err := index.GetFromIndex[corev1.Node](ctx, runner.sharedIndexInformer, groups.SchedulingGroupIdx, string(key))
+	if err != nil {
+		return nil, err
+	}
+
+	candidates := make([]*corev1.Node, 0)
+	for _, node := range nodes {
+		taint, exist := k8sclient.GetTaint(node)
+		// if the node doesn't have the draino taint, it should not be processed
+		if !exist {
+			continue
+		}
+		// if the taint value is "drained", draino is done and the node should be ignored
+		if taint.Value == k8sclient.TaintDrained {
```

also `TaintDraining` ?
Finally why not `taint.Value!=TainDrainCandidate` ?

---

As discussed, we'll only care about drain candidates and ignore error states here. The cleanup of nodes in invalid states will be done somewhere else.